test_quality_of_service/test/test_liveliness.cpp

```diff
@@ -0,0 +1,158 @@
+#include <exception>
+#include <iostream>
+#include <memory>
+#include <stack>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+using namespace std::chrono_literals;
+
+/// Liveliness setup
+class LivelinessSetup : public ::testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+    executor = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+  }
+  void TearDown() override
+  {
+    executor->cancel();
+    executor.reset();
+    rclcpp::shutdown();
+  }
+
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> executor;
+};
+
+/// Test Automatic Liveliness with a single publishing node and single subscriber node
+TEST_F(LivelinessSetup, test_automatic_liveliness_changed) {
+  const std::chrono::milliseconds max_test_length = 8s;
+  const std::chrono::milliseconds kill_publisher_after = 2s;
+  int number_of_published_messages = 0;
+
+  // used for lambda capture
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec = executor;
+  int total_number_of_liveliness_events = 0;
+
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_default;
+  // define liveliness policy
+  qos_profile.liveliness = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
+
+  std::tuple<size_t, size_t> liveliness_lease_tuple = chrono_milliseconds_to_size_t(1000ms);
+
+  // setup liveliness lease
+  qos_profile.liveliness_lease_duration.sec = std::get<0>(liveliness_lease_tuple);
+  qos_profile.liveliness_lease_duration.nsec = std::get<1>(liveliness_lease_tuple);
+
+  // subscription options
+  rclcpp::SubscriptionOptions<> subscriber_options;
+  subscriber_options.qos_profile = qos_profile;
+
+  // subscriber Liveliness callback event
+  subscriber_options.event_callbacks.liveliness_callback =
+    [&total_number_of_liveliness_events](rclcpp::QOSLivelinessChangedInfo & event) -> void
+    {
+      RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"), "QOSLivelinessChangedInfo callback");
+      total_number_of_liveliness_events++;
+
+      // strict checking for expected events
+      if (total_number_of_liveliness_events == 1) {
+        // publisher came alive
+        ASSERT_EQ(1, event.alive_count);
+        ASSERT_EQ(0, event.not_alive_count);
+        ASSERT_EQ(1, event.alive_count_change);
+        ASSERT_EQ(0, event.not_alive_count_change);
+      } else if (total_number_of_liveliness_events == 2) {
+        // publisher died
+        ASSERT_EQ(0, event.alive_count);
+        ASSERT_EQ(0, event.not_alive_count);
+        ASSERT_EQ(-1, event.alive_count_change);
+        ASSERT_EQ(0, event.not_alive_count_change);
+      }
+    };
+
+  // publisher options
+  rclcpp::PublisherOptions<> publisher_options;
+  publisher_options.qos_profile = qos_profile;
+
+//  publisher_options.event_callbacks.liveliness_callback =
```

Comment:
Don't leave any commented code

Reply:
Actually I left it commented as a question for you (so great catch reminding me :-)), @nburek , and @mm318 : I can't get this callback to fire when uncommented. Expected?